include/openthread.h

```diff
@@ -995,6 +1061,46 @@ int otWriteMessage(otMessage aMessage, uint16_t aOffset, const void *aBuf, uint1
  */
 
 /**
+ * @addtogroup raw-ip6  Raw IPv6
+ *
+ * @brief
+ *   This module includes functions that control raw IPv6 communication.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This function pointer is called when an IPv6 datagram is received.
+ *
+ * @param[in]  aMessage  A pointer to the message buffer containing the received IPv6 datagram.
+ *
+ */
+typedef void (*otRawIp6ReceivedCallback)(otMessage aMessage);
+
+/**
+ * This function registers a callback to provide receivd raw IPv6 datagrams.
+ *
+ * @param[in]  aCallback  A pointer to a function that is called when an IPv6 datagram is received or NULL to disable
+ *                        the callback.
+ *
+ */
+void otSetRawIp6ReceivedCallback(otRawIp6ReceivedCallback aCallback);
```

Comment:
Regarding callbacks... If registering for callbacks is how callbacks are going to work (as opposed to declaring specific callback functions that the client implements directly), then I think having a context pointer is important. Without it you have to resort to hacks using global variables (see how I handled the scan result callback, for example).